### PR TITLE
Allow is_unknown_time to work on arrays

### LINKED ISSUE #814: is_unknown_time for numpy arrays
The current `is_unknown_time()` function only works one number at a time. It would be nice to be able to do

```
def is_unknown_time_array(a):
    np_unknown_time = np.float64(tskit.UNKNOWN_TIME)
    return a.view(np.uint64) == np_unknown_time.view(np.uint64)
```

e.g.

```
>>> a = np.full(10, tskit.UNKNOWN_TIME)
>>> np.all(is_unknown_time_array(a))
True
```

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #814

### CODE REVIEW

**Comment by hyanwong** [comment]
LGTM!

**Review by hyanwong — APPROVED** [review]

File: docs/data-model.rst [hunk 1]
```diff
@@ -634,12 +634,12 @@ requirements for a valid set of mutations are:
...
 - ``node`` must refer to a valid node ID;
-- ``time`` must either be UNKNOWN_TIME (a NAN value which indicates
+- ``time`` must either be ``UNKNOWN_TIME`` (a NAN value which indicates
   the time is unknown) or be a finite value which is greater or equal to the
   mutation ``node``'s ``time``, less than the ``node`` above the mutation's
   ``time`` and equal to or less than the ``time`` of the ``parent`` mutation
-  if this mutation has one. If one mutation on a site has UNKNOWN_TIME then all mutations
-  at that site must, a mixture of known and unknown is not valid.
+  if this mutation has one. If one mutation on a site has ``UNKNOWN_TIME`` then
+  all mutations at that site must, a mixture of known and unknown is not valid.
```
Comment (jeromekelleher) [inline docs/data-model.rst:642]:
```suggestion
  all mutations at that site must, as a mixture of known and unknown is not valid.
```

File: docs/data-model.rst [hunk 2]
```diff
@@ -668,6 +668,9 @@ mutation does not result in any change of state. This error is
 raised at run-time when we reconstruct sample genotypes, for example
 in the :meth:`TreeSequence.variants` iterator.
 
+.. note:: As ``tskit.UNKNOWN_TIME`` is implemented as a ``NaN`` value, tests for
+    equality will always fail. Use ``util.is_unknown_time`` to detect unknown
```
Comment (jeromekelleher) [inline docs/data-model.rst:672]: We don't expose the package structure - just ``tskit.is_unknown_time`` presumably?

File: python/CHANGELOG.rst [hunk 3]
```diff
@@ -28,6 +28,8 @@
 
 - New ``tree.is_isolated(u)`` method (:user:`hyanwong`, :pr:`443`).
 
+- ``util.is_unknown_time`` can now check arrays. (:user:`benjeffery`, :pr:`857`).
```
Comment (jeromekelleher) [inline python/CHANGELOG.rst:31]: Ditto utuil here

File: python/tskit/util.py [hunk 4]
```diff
@@ -45,9 +45,18 @@ def canonical_json(obj):
...
+    method compares the bitfield such that unknown times can be detected.
+    Either single floats can be passed or lists/arrays.
+
+    :param float or array-like time: Value or array to check.
+    :return: A single boolean or array of booleans the same shape as ``time``.
+    :rtype: bool or np.array(dtype=bool)
     """
-    return struct.pack(">d", UNKNOWN_TIME) == struct.pack(">d", time)
+    try:
+        return struct.pack(">d", UNKNOWN_TIME) == struct.pack(">d", time)
```
Comment (jeromekelleher) [inline python/tskit/util.py:56]: Do we need to first part now? I thought numpy would cast to a 1-element array.

**Review by jeromekelleher — COMMENTED** [review]
LGTM, minor comments.

File: docs/data-model.rst (hunk 1 above)
Comment (benjeffery) [reply docs/data-model.rst]: Done

File: docs/data-model.rst (hunk 2 above)
Comment (benjeffery) [reply docs/data-model.rst]: Opps, keep tripping over this one. Fixed.

File: python/CHANGELOG.rst (hunk 3 above)
Comment (benjeffery) [reply python/CHANGELOG.rst]: Fixed

File: python/tskit/util.py (hunk 4 above)
Comment (benjeffery) [reply python/tskit/util.py]: Seems good - I get a `bool` back! Nice.

**Comment by benjeffery** [comment]
@jeromekelleher All fixed up.

**Comment by jeromekelleher** [comment]
Issues on Windows?

**Comment by benjeffery** [comment]
Huh, weird. Looking into it.

**Comment by benjeffery** [comment]
@jeromekelleher Fixed by forcing the dtype of the array, means that `None` now gives false instead of an error, but it feels more correct to do that anyway.

File: python/tskit/util.py [hunk 5]
```diff
@@ -45,9 +44,20 @@ def canonical_json(obj):
...
-    method compares the bitfield.
+    method compares the bitfield such that unknown times can be detected.
+    Either single floats can be passed or lists/arrays.
+
+    :param float or array-like time: Value or array to check.
+    :return: A single boolean or array of booleans the same shape as ``time``.
+    :rtype: bool or np.array(dtype=bool)
     """
-    return struct.pack(">d", UNKNOWN_TIME) == struct.pack(">d", time)
+    # try:
```
Comment (jeromekelleher) [inline python/tskit/util.py:54]: Is there a reason for leaving the commented out code in here?

**Comment by benjeffery** [comment]
@jeromekelleher Fixed my pub-induced error.

**Review by jeromekelleher — APPROVED** [review]

**Pull request merged**